### PR TITLE
Post new alerts on IthacaTransit Twitter

### PR DESCRIPTION
**Pull request by alanna-zhou**
Issue #11 tested and works on Twitter

### CODE REVIEW

File: src/alerts.py [hunk 1]
```diff
@@ -42,21 +60,24 @@ def fetch_alerts(event):
...
+          'fromTime': format_date(alert_dict.get('FromTime')),
           'id': alert_dict.get('MessageId'),
           'message': alert_dict.get('Message'),
           'priority': alert_dict.get('Priority'),
           'routes': alert_dict.get('Routes'),
           'signs': alert_dict.get('Signs'),
-          'toDate': convert_to_date(alert_dict.get('ToDate')),
-          'toTime': convert_to_date(alert_dict.get('ToTime'))
+          'toDate': format_date(alert_dict.get('ToDate')),
+          'toTime': format_date(alert_dict.get('ToTime'))
```
Comment (young-k) [inline src/alerts.py]: aren't these keys guaranteed? can we just do `[]` and have an `try` `except` clause instead?

File: src/twitter.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import os
+import tweepy
+
+TWITTER_SCREEN_NAME = 'IthacaTransit'
+NUM_RECENT_STATUSES = 100
+
+twitter_api = None
+
+def set_twitter_api():
```
Comment (young-k) [inline src/twitter.py]: this should probably be a `TwitterAPI` class, so we avoid using `global`, and the `set_twitter_api` method can be inside of the constructor

File: src/twitter.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  """Filter the array of statuses by removing already tweeted statuses"""
+  for tweet in get_tweets():
+    if tweet.text in statuses:
+      statuses.remove(tweet.text)
+  return statuses
+
+def tweet(statuses):
+  """Makes a request to post a tweet with the given status"""
+  global twitter_api
+  if twitter_api is None:
```
Comment (young-k) [inline src/twitter.py]: Same comment as before

File: src/twitter.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    set_twitter_api()
+  return twitter_api.user_timeline(screen_name=TWITTER_SCREEN_NAME, count=NUM_RECENT_STATUSES)
+
+def get_new_statuses(statuses):
+  """Filter the array of statuses by removing already tweeted statuses"""
+  for tweet in get_tweets():
+    if tweet.text in statuses:
+      statuses.remove(tweet.text)
```
Comment (young-k) [inline src/twitter.py]: Maybe cleaner, @meganle what are your thoughts:
```
tweet_texts = [t.text for t in get_tweets()]
return [s for s in statuses if s not in tweet_texts]
```

File: src/twitter.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  global twitter_api
+  auth = tweepy.OAuthHandler(os.environ['TWITTER_KEY'], os.environ['TWITTER_KEY_SECRET'])
+  auth.set_access_token(os.environ['TWITTER_TOKEN'], os.environ['TWITTER_TOKEN_SECRET'])
+  twitter_api = tweepy.API(auth)
+
+def get_tweets():
+  """Gets the last n status objects from a twitter user's timeline"""
+  global twitter_api
+  if twitter_api is None:
+    set_twitter_api()
```
Comment (young-k) [inline src/twitter.py]: this check can be removed post-class conversion, also you can just write: `if not twitter_api:`

File: src/alerts.py [hunk 6]
```diff
@@ -32,8 +35,23 @@ def convert_num_to_str(tcat_num):
   }
   return num_day_dict.get(tcat_num, '')
 
+def convert_to_date(date_str):
```
Comment (young-k) [inline src/alerts.py]: `convert_str_to_date`?

File: src/alerts.py [hunk 7]
```diff
@@ -32,8 +35,23 @@ def convert_num_to_str(tcat_num):
...
+  return datetime.strptime(date_str, DATE_STRING_FORMAT)
+
+def format_alert_status(alert):
+  """Create a string status from an alert's message and an Effective description to tweet."""
```
Comment (yuna878) [inline src/alerts.py]: Is the capitalization in `Effective` a typo or is it a thing?

File: src/twitter.py (hunk 4 above)
Comment (meganle) [reply src/twitter.py]: wow "very nice"

File: src/alerts.py (hunk 1 above)
Comment (kevinchan159) [reply src/alerts.py]: is using `get` bad? I think its better to return some data rather than none if the key doesn't exist

Comment (young-k) [reply src/alerts.py]: discussed offline

File: src/twitter.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+      effective_str = f"Effective: {from_date_str} - {to_date_str}"
+    return f"{alert['message']}\n{effective_str}"
+
+  def get_twitter_api(self):
+    """Authenticates with client tokens and returns tweepy api object"""
+    auth = tweepy.OAuthHandler(os.environ['TWITTER_KEY'], os.environ['TWITTER_KEY_SECRET'])
+    auth.set_access_token(os.environ['TWITTER_TOKEN'], os.environ['TWITTER_TOKEN_SECRET'])
+    return tweepy.API(auth)
+
+  def get_statuses(self):
```
Comment (young-k) [inline src/twitter.py]: i think it makes more sense to have:

```
def get_statuses(self, n):
"""Gets the last {n} status messages from a user's timeline"""
```

File: src/twitter.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
+from calendar import month_abbr
+from datetime import datetime
+import html
+import os
+
+import tweepy
+
+TWITTER_SCREEN_NAME = 'IthacaTransit'
+NUM_RECENT_STATUSES = 100
+DATE_STRING_FORMAT = '%Y-%m-%dT%H:%M:%S.%fZ'
```
Comment (young-k) [inline src/twitter.py]: i think all of these constants should be moved within the class, since nothing else is using them

File: src/twitter.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    to_date = self.convert_str_to_date(alert['toDate'])
+    from_date_str = f"{month_abbr[from_date.month]} {from_date.day}"
+    if from_date == to_date:
+      effective_str = f"Effective: {from_date_str}"
+    else:
+      to_date_str = f"{month_abbr[to_date.month]} {to_date.day}"
+      effective_str = f"Effective: {from_date_str} - {to_date_str}"
+    return f"{alert['message']}\n{effective_str}"
+
+  def get_twitter_api(self):
```
Comment (young-k) [inline src/twitter.py]: why isn't this function just moved into the `init`? nothing else is calling it, and the `init` method doesn't have anything to justify moving this into a separate function

File: src/twitter.py (hunk 9 above)
Comment (young-k) [reply src/twitter.py]: not sorted @alanna-zhou 😠 i thought you sang the alphabet
cc @meganle

**Pull request merged**